packages/editor/src/components/convert-to-group-buttons/convert-button.js

```diff
@@ -0,0 +1,130 @@
+/**
+ * External dependencies
+ */
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { Fragment } from '@wordpress/element';
+import { MenuItem } from '@wordpress/components';
+import { _x } from '@wordpress/i18n';
+import { switchToBlockType } from '@wordpress/blocks';
+import { withSelect, withDispatch } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+import { displayShortcut } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import { Group, UnGroup } from './icons';
+
+export function ConvertToGroupButton( {
+	onConvertToGroup,
+	onUnCovertFromGroup,
+	isGroupable = false,
+	isUnGroupable = false,
+} ) {
+	return (
+		<Fragment>
+			{ isGroupable && (
+				<MenuItem
+					className="editor-block-settings-menu__control block-editor-block-settings-menu__control"
+					icon={ Group }
+					onClick={ onConvertToGroup }
+					shortcut={ displayShortcut.primaryAlt( 'g' ) }
+				>
+					{ _x( 'Group', 'Grouping a selection of blocks together within the Editor' ) }
+				</MenuItem>
+			) }
+			{ isUnGroupable && (
+				<MenuItem
+					className="editor-block-settings-menu__control block-editor-block-settings-menu__control"
+					icon={ UnGroup }
+					onClick={ onUnCovertFromGroup }
+					shortcut={ displayShortcut.secondary( 'g' ) }
+				>
+					{ _x( 'Ungroup', 'Ungrouping blocks from within a Group block back into individual blocks within the Editor ' ) }
+				</MenuItem>
+			) }
+		</Fragment>
+	);
+}
+
+export default compose( [
+	withSelect( ( select, { clientIds } ) => {
+		const {
+			getBlocksByClientId,
+			canInsertBlockType,
+		} = select( 'core/block-editor' );
+
+		const containerBlockAvailable = canInsertBlockType( 'core/group' );
+
+		const blocksSelection = getBlocksByClientId( clientIds );
+
+		const isSingleContainerBlock = blocksSelection.length === 1 && blocksSelection[ 0 ] && blocksSelection[ 0 ].name === 'core/group';
+
+		// Do we have
+		// 1. Container block available to be inserted?
+		// 2. One or more blocks selected
+		// (we allow single Blocks to become groups unless
+		// they are a soltiary group block themselves)
+		const isGroupable = (
+			containerBlockAvailable &&
+			blocksSelection.length &&
+			! isSingleContainerBlock
+		);
+
+		// Do we have a single Group Block selected?
+		const isUnGroupable = isSingleContainerBlock;
+
+		return {
+			isGroupable,
+			isUnGroupable,
+			blocksSelection,
```

Comment:
If you're only using this for `withDispatch`, I'd recommend using the third argument of the `withDispatch` callback to access the current registry. Then, you can call `select` only if and when relevant. It's particularly important here for `getBlocksByClientId` which is not a very performant selector.  But it's a good pattern to follow in general.

See also:

- https://github.com/WordPress/gutenberg/pull/15499#discussion_r282128324 (from feedback on your earlier pull request https://github.com/WordPress/gutenberg/pull/13899#discussion_r266877788 )
- #11851

Reply:
All the selectors I'm using are required to determine whether the Group/Ungroup interactions should be available and visible in the UI. There was a point where I was relying on the select solely to pass data into dispatch but I don't think I'm doing that anymore (am I?).

I wasn't aware of the 3rd arg to `withDispatch` so I'll try to be mindful of that in future. Much appreciated.

Reply:
> All the selectors I'm using are required to determine whether the Group/Ungroup interactions should be available and visible in the UI. There was a point where I was relying on the select solely to pass data into dispatch but I don't think I'm doing that anymore (am I?).

You're right. I was only looking at the component itself, which did not use `blocksSelection`. Since it's needed within the `mapSelectToProps` callback itself, it should be fine as-is.

Reply:
Thanks @aduth.